### PR TITLE
adds extractors for TypeName, TermName and Modifiers 

### PR DESCRIPTION
**Pull request by xeno-by**

### CODE REVIEW

**Comment by xeno-by** [comment]
Resubmission of https://github.com/scala/scala/pull/1753 against `master`

**Comment by xeno-by** [comment]
review @paulp @retronym

**Comment by xeno-by** [comment]
PLS REBUILD ALL

File: test/files/run/reflection-enclosed-basic.scala [hunk 1]
```diff
@@ -20,7 +20,7 @@ object Test extends App {
   def testNestedClass(name: String) = {
     val sym = cm.staticClass(name)
     println(sym)
-    val ctor = sym.typeSignature.declaration(newTermName("<init>")).asMethod
+    val ctor = sym.typeSignature.declaration(TermName("<init>")).asMethod
```
Comment (paulp) [inline test/files/run/reflection-enclosed-basic.scala]: I know it isn't really this commit's problem, but this should certainly be nme.CONSTRUCTOR.

**Comment by paulp** [comment]
lgtm.

File: test/files/run/reflection-enclosed-basic.scala (hunk 1 above)
Comment (xeno-by) [inline test/files/run/reflection-enclosed-basic.scala]: Done

**Comment by xeno-by** [comment]
PLS REBUILD ALL

**Pull request merged**